handlers/src/main/java/com/nextdoor/bender/handler/kinesis/KinesisEventIterator.java

```diff
@@ -54,7 +69,19 @@ public boolean hasNext() {
 
   @Override
   public InternalEvent next() {
-    return new KinesisInternalEvent(this.iterator.next(), this.context, this.shardId);
+    KinesisEventRecord nextRecord = this.iterator.next();
+    if (assumeKinesisDataIsGzipped) {
+      ByteBuffer gzip = nextRecord.getKinesis().getData();
+      try (GZIPInputStream gzipInputStream = new GZIPInputStream(new ByteArrayInputStream(gzip.array()))) {
+        IOUtils.copy(gzipInputStream, byteArrayOutputStream, bufferSize);
```

Comment:
Experiment and see if you can use piped streams instead of doing a copy. Same thing below, would like to limit the number of byte objects we're creating to store the same data just to move it between classes. Eg seeing if there is a way to wrap an output stream via a ByteBuffer.

Reply:
Synced offline with Slava on this one, doing a workflow like `gzipstream -> piped output -> piped input -> byte stream that backs a byte buffer` is a bit tricky since the BaseHandler iterates through all the internal events first, so we'd need to add a break ourselves so that the deserializer reads one event at a time. We're going to stay off this for now as a result.